File: packages/ios/Sources/Components/Input/WashedField.swift
import SwiftUI

/// A field washed rather than outlined, as React draws the route points and
/// the feedback box: no edge, black at 5 % on light and white at 10 % on dark,
/// doubled while focused, the control radius, and the standard focus ring —
/// 2 pt in the ring role, 2 pt out. `multiline` makes it a text area, 80 high
/// at least and growing with its text, as React's is.
struct KozmosWashedField: View {
    @Environment(\.colorScheme) private var colorScheme
    @Binding var text: String
    let placeholder: String
    var multiline: Bool = false
    @FocusState private var focused: Bool

    var body: some View {
        let radius = KozmosDimensions.semanticsRadiusControl
        let wash = (colorScheme == .dark ? 0.10 : 0.05) * (focused ? 2 : 1)
        let prompt = Text(placeholder).foregroundColor(KozmosColors.primitivesColorsForeground400)
        Group {
            if multiline {
                TextField("", text: $text, prompt: prompt, axis: .vertical)
                    .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing100)
                    .frame(minHeight: 80, alignment: .topLeading)
            } else {
                TextField("", text: $text, prompt: prompt)
                    .frame(height: 40)
            }
        }
        .textFieldStyle(.plain)
        .font(KozmosTypography.subheadline)
        .foregroundColor(KozmosColors.primitivesColorsForeground0)
        .tint(KozmosColors.primitivesColorsTheme600)
        .focused($focused)
        .accessibilityLabel(placeholder)
        .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing150)
        .background(
            RoundedRectangle(cornerRadius: radius, style: .continuous)
                .fill(KozmosColors.primitivesColorsForeground0.opacity(wash))
        )
        .overlay(
            RoundedRectangle(cornerRadius: radius + 3, style: .continuous)
                .stroke(KozmosColors.primitivesColorsTheme600, lineWidth: 2)
                .padding(-3)
                .opacity(focused ? 1 : 0)
        )
    }
}

File: packages/android/src/main/java/com/kozmos/components/Input/WashedField.kt
package com.kozmos.components.input

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/**
 * A field washed rather than outlined, as React draws the route points and the
 * feedback box: no edge, black at 5 % on light and white at 10 % on dark,
 * doubled while focused, the control radius, and the ring role's 2 dp outline
 * marking focus. Single-line it is 40 high, 12 in; [multiline] makes it a text
 * area, 80 high at least, 8 and 12 in, as React's is.
 */
@Composable
internal fun KozmosWashedField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    modifier: Modifier = Modifier,
    multiline: Boolean = false
) {
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
    val interaction = remember { MutableInteractionSource() }
    val focused by interaction.collectIsFocusedAsState()
    val wash = (if (KozmosThemeTokens.isDark) 0.10f else 0.05f) * (if (focused) 2f else 1f)
    val ring = KozmosThemeTokens.primitivesColorsTheme600
    val textStyle = MaterialTheme.typography.bodyMedium
    val placeholderColor = KozmosThemeTokens.primitivesColorsForeground400
    BasicTextField(
        value = value,
        onValueChange = onValueChange,
        singleLine = !multiline,
        textStyle = textStyle.copy(color = KozmosThemeTokens.primitivesColorsForeground0),
        cursorBrush = SolidColor(ring),
        interactionSource = interaction,
        modifier = modifier
            .then(if (multiline) Modifier.heightIn(min = 80.dp) else Modifier.height(40.dp))
            .background(KozmosThemeTokens.primitivesColorsForeground0.copy(alpha = wash), shape)
            .then(if (focused) Modifier.border(2.dp, ring, shape) else Modifier)
            .semantics { contentDescription = placeholder },
        decorationBox = { inner ->
            Box(
                contentAlignment = if (multiline) Alignment.TopStart else Alignment.CenterStart,
                modifier = Modifier
                    .then(if (multiline) Modifier else Modifier.fillMaxSize())
                    .padding(horizontal = 12.dp, vertical = if (multiline) 8.dp else 0.dp)
            ) {
                if (value.isEmpty()) {
                    Text(placeholder, style = textStyle, color = placeholderColor, maxLines = if (multiline) 3 else 1)
                }
                inner()
            }
        }
    )
}
